contrib/go-redis/redis/redis.go

```diff
@@ -139,29 +134,25 @@ func commandsToString(cmds []redis.Cmder) string {
 
 // WithContext sets a context on a Client. Use it to ensure that emitted spans have the correct parent.
 func (c *Client) WithContext(ctx context.Context) *Client {
-	c.mu.Lock()
-	c.ctx = ctx
-	c.mu.Unlock()
-	return c
-}
-
-// Context returns the active context in the client.
-func (c *Client) Context() context.Context {
-	c.mu.RLock()
-	ctx := c.ctx
-	c.mu.RUnlock()
-	return ctx
+	clone := &Client{
+		Client:  c.Client.WithContext(ctx),
+		params:  c.params,
+		process: c.process,
+	}
+	clone.Client.WrapProcess(createWrapperFromClient(clone))
+	return clone
 }
 
 // createWrapperFromClient returns a new createWrapper function which wraps the processor with tracing
 // information obtained from the provided Client. To understand this functionality better see the
 // documentation for the github.com/go-redis/redis.(*baseClient).WrapProcess function.
 func createWrapperFromClient(tc *Client) func(oldProcess func(cmd redis.Cmder) error) func(cmd redis.Cmder) error {
 	return func(oldProcess func(cmd redis.Cmder) error) func(cmd redis.Cmder) error {
+		if tc.process == nil {
+			tc.process = oldProcess
```

Comment:
I’m afraid this could lead to a race condition. Better simply check if it’s set and use it, and if not just use oldProcess. It’s more verbose, but more clear, obvious, and race free

Reply:
We need to set it somewhere, and this is the only place we get access to `oldProcess`.

I don't think there can be a race here. `createWrapperFromClient` is not exported, and it's only ever called and passed to `WrapProcess` once per `Client` - in `WrapClient` when the client is being created, or in `WithContext` where it's called on the clone. It's not possible for this function to be run on a `Client` more than once, with or without multiple goroutines. Once a pointer to a `Client` is returned, `createWrapperFromClient` can't be called on it.

There's also no race between creating a client and cloning it with `WithContext`. In both cases where a `Client` is created (`WrapClient` and `WithContext`), the wrapper is set before a reference to the client is returned. Nobody can have a pointer to the `Client` before it is set up, and `process` can't be altered later. 

To be clear, this check doesn't run every time the process function is called, this is the function that is called to wrap the process function. The logic only runs once when setting up the `Client` when we call `WrapProcess`. At that point, this conditional is executed, and the innermost `func` here is returned to be the new `process` function.

Please let me know if I've missed something.

Reply:
It was me who missed something. Sorry about that. I only now realised this is happening outside the actual process function. Resolving.